### PR TITLE
Platform channel for predictive back in route transitions on android

### PR DESCRIPTION
**Pull request by maRci002**
This pull request introduces platform channel support for predictive back route transitions on Android, as part of the ongoing efforts in flutter/flutter#131961.

- For Android SDK 33: Utilizes [`OnBackInvokedCallback`](https://developer.android.com/reference/android/window/OnBackInvokedCallback) to handle back navigation events (#39208).
- For Android SDK 34 and above: Employs [`OnBackAnimationCallback`](https://developer.android.com/reference/android/window/OnBackAnimationCallback), taking advantage of the enhanced back navigation capabilities introduced in this version.

In both scenarios, the callbacks are appropriately forwarded to the Flutter framework.
This PR focuses on the engine part, setting the stage for integrating these enhancements into the broader Flutter ecosystem.

### References

Part of flutter/flutter#131961
Framework PR: flutter/flutter#141373
Fixes https://github.com/flutter/flutter/issues/111295

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: shell/platform/android/io/flutter/embedding/engine/systemchannels/NavigationChannel.java
Comment (camsim99) [inline shell/platform/android/io/flutter/embedding/engine/systemchannels/NavigationChannel.java:95]: Nit: You could just forward the Android constants to the framework and decode it on that side with a comment linking to the docs instead of using these Strings for clarity.

File: shell/platform/android/io/flutter/embedding/android/FlutterActivity.java
Comment (camsim99) [inline shell/platform/android/io/flutter/embedding/android/FlutterActivity.java:662]: Do we need both of these checks?

File: shell/platform/android/io/flutter/embedding/android/FlutterActivity.java
Comment (camsim99) [inline shell/platform/android/io/flutter/embedding/android/FlutterActivity.java:684]: It may be nice to write a test that verifies this callback responds to each method as we'd expect.

**Review by camsim99 — COMMENTED** [review]
This LGTM overall just left an idea for a test and some nits!

**Review by justinmc — COMMENTED** [review]
@maRci002 You are a hero for implementing this, thank you for the PR.

Have you tried tying into these methods on the framework side and seeing what happens? I'd like to verify that it is going to meet our needs on the framework before merging this PR. I added some requirements to the description in https://github.com/flutter/flutter/issues/131961 that I'd like to verify are possible.

You mentioned in https://github.com/flutter/flutter/issues/131961#issuecomment-1865076080 that you might want to work on the framework side. Feel free to open a framework PR (even a draft) if so. We don't have to hit all of those requirements in the first PR, I just want to make sure we've got the API right so that they're possible.

File: shell/platform/android/io/flutter/embedding/android/FlutterActivity.java
Comment (maRci002) [reply shell/platform/android/io/flutter/embedding/android/FlutterActivity.java:662]: Unfortunately, yes, since the compiler does not follow what's going on inside the `createOnBackInvokedCallback` method.

I can remove the method and use a nested Elvis operator:
```java
  final OnBackInvokedCallback onBackInvokedCallback =
          Build.VERSION.SDK_INT == 33 ? this::onBackPressed : (Build.VERSION.SDK_INT >= 34 ?
                  new OnBackAnimationCallback() {
                    @Override
                    public void onBackInvoked() {
                      commitBackGesture();
                    }

                    @Override
                    public void onBackCancelled() {
                      cancelBackGesture();
                    }

                    @Override
                    public void onBackProgressed(@NonNull BackEvent backEvent) {
                      updateBackGestureProgress(backEvent);
                    }

                    @Override
                    public void onBackStarted(@NonNull BackEvent backEvent) {
                      startBackGesture(backEvent);
                    }
                  } : null);
```

Or mark `createOnBackInvokedCallback` return value as `@NonNull`:
```java
  final OnBackInvokedCallback onBackInvokedCallback = Build.VERSION.SDK_INT < 33 ? null : createOnBackInvokedCallback();

  @NonNull
  private OnBackInvokedCallback createOnBackInvokedCallback() {
    if (Build.VERSION.SDK_INT >= 34) {
      return new OnBackAnimationCallback() {
        @Override
        public void onBackInvoked() {
          commitBackGesture();
        }

        @Override
        public void onBackCancelled() {
          cancelBackGesture();
        }

        @Override
        public void onBackProgressed(@NonNull BackEvent backEvent) {
          updateBackGestureProgress(backEvent);
        }

        @Override
        public void onBackStarted(@NonNull BackEvent backEvent) {
          startBackGesture(backEvent);
        }
      };
    }

    return this::onBackPressed;
  }
```

File: shell/platform/android/io/flutter/embedding/android/FlutterActivity.java
Comment (maRci002) [reply shell/platform/android/io/flutter/embedding/android/FlutterActivity.java]: done

**Comment by maRci002** [comment]
Thanks, @justinmc. I've just created the PR for the framework part flutter/flutter#141373

File: shell/platform/android/io/flutter/embedding/android/FlutterActivity.java
Comment (camsim99) [reply shell/platform/android/io/flutter/embedding/android/FlutterActivity.java]: I don't believe that warning should cause that failure. We use compileSdk version 34 but the warning is just suggestion we upgrade our AGP version to accommodate that fact.

To get the test running against API 34, you need to annotate the test with `@Config(sdk = 34)`. Can you try making that change, then checking the type of `onBackInvokedCallback`?

Comment (maRci002) [reply shell/platform/android/io/flutter/embedding/android/FlutterActivity.java]: I've attempted to use the `@Config(sdk = 34)` annotation to test against API 34, but I encountered problems. In the meantime, I have created separate tests to cover different SDK versions and scenarios: `itHandlesOnBackAnimationCallbackAsExpectedForSdk34OrHigher`, `onBackInvokedCallbackCallsOnBackPressedForSdk33`, and `onBackInvokedCallbackIsNullForSdk32OrLower`.

Could you please take a look at the modifications I've made to the tests?

**Comment by jonahwilliams** [comment]
What is the status of this PR, is it waiting on review comments or framework changes? Friendly ping @justinmc / @maRci002

**Comment by maRci002** [comment]
In my opinion, this PR is ready for further review. However, I've encountered some issues with testing, particularly related to the `@Config` annotation. I'm currently awaiting some inputs from @camsim99 to resolve these concerns. Once these are addressed, I believe the PR will be in good shape for final review.

**Comment by justinmc** [comment]
I owe a major review on this, sorry for the delay. I've been struggling to get my engine development set up on a new machine...

File: shell/platform/android/io/flutter/embedding/android/FlutterActivity.java
Comment (camsim99) [reply shell/platform/android/io/flutter/embedding/android/FlutterActivity.java]: The tests look good to me! What issues are you running into?

Looks like the 34 test may be failing because of the verifications for the callback in the test. I'm not familiar with `FlutterActivityWithReportFullyDrawn` so it may have to do with its implementation if you've double checked those are the method calls we need to verify there.

**Comment by camsim99** [comment]
@maRci002 Looks like tests are passing! Is this ready for re-review?

File: shell/platform/android/io/flutter/embedding/android/FlutterActivity.java
Comment (justinmc) [inline shell/platform/android/io/flutter/embedding/android/FlutterActivity.java:696]: So for API >= 34, popping will be done by the framework via commitBackGesture. For API <= 33, the framework will handle it by listening to popRoute called by onBackPressed as usual.

Could there be any situations where users that are expecting onBackPressed won't get it and won't be listening for commitBackGesture? Like what if someone has a custom Android route transition and they don't know about these new changes?

File: shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java
Comment (justinmc) [inline shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java:802]: I'm debating if we should create a new channel for this. This looks independent of everything else in the navigation channel, even though it does relate to navigation.

I'll comment more about this on the framework PR.

**Review by justinmc — COMMENTED** [review]
Sorry for such a long delay in reviewing this. I was able to get it running locally and it worked great for me! I think overall this approach looks good.

My main concern is any possible breakages due to no longer calling onBackPressed. See my comment on that below.

I'll also leave a review on the framework PR in a bit.

File: shell/platform/android/io/flutter/embedding/android/FlutterActivity.java
Comment (justinmc) [reply shell/platform/android/io/flutter/embedding/android/FlutterActivity.java:696]: Ok at the route transition level it won't matter because either way it will just result in a pop. But I wonder if this could mess with users at any other level...

Comment (maRci002) [reply shell/platform/android/io/flutter/embedding/android/FlutterActivity.java:696]: > But I wonder if this could mess with users at any other level...

If someone directly sets their own method for `SystemChannels.navigation.setMethodCallHandler`, it might cause a problem. However, using the recommended `WidgetsBindingObserver` to register their own handlers shouldn't be a problem, since the `handleCommitBackGesture` method might delegate work to the `handlePopRoute` method if it's not handled by registered observers.

**Comment by maRci002** [comment]
> @maRci002 Looks like tests are passing! Is this ready for re-review?

@camsim99, thanks, but the API might change over time. I will definitely let you know when it's ready for review from an Android perspective.

File: shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java
Comment (maRci002) [reply shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java]: I have added `BackGestureChannel`

**Comment by justinmc** [comment]
@maRci002 Can you fix the build conflict? I think it's just that PR https://github.com/flutter/engine/pull/51171 requires the use of `API_LEVELS.API_33` instead of `33` etc. now.

**Review by justinmc — APPROVED** [review]
LGTM with the new method channel :+1: . Thanks for adding that.

I think once you've fixed the merge conflict we should move forward with merging this PR, and then focus on getting the framework PR merged.

**Comment by maRci002** [comment]
> I think once you've fixed the merge conflict we should move forward with merging this PR, and then focus on getting the framework PR merged.

Framework and Engine should be merged together; otherwise, on Android 14, the back button will break down since the `commitBackGesture` event will be sent instead of `popRoute`.

@camsim99, thank you for waiting. The PR is ready for review from an Android perspective. However, please do not apply `autosubmit` label if possible until the framework is finished. I will let you know when it's ready :)

**Comment by justinmc** [comment]
Ah good catch. Sounds good.

File: shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java
Comment (camsim99) [inline shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java:791]: Nit: can this doc is some way clarify that what it's responsible for is defined in the framework?

Comment (camsim99) [reply shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java:791]: Same nit for the other methods you added here

**Review by camsim99 — APPROVED** [review]
Looks great! Just left an optional nit on documentation

File: shell/platform/android/io/flutter/embedding/engine/systemchannels/BackGestureChannel.java
Comment (justinmc) [inline shell/platform/android/io/flutter/embedding/engine/systemchannels/BackGestureChannel.java:40]: @maRci002 I'm noticing that most platform channel names are all lowercase. Can you change this to "flutter/backgesture"?

**Comment by justinmc** [comment]
Thanks for the quick turnaround on that!

**Comment by timbotimbo** [comment]
Any idea how these changes can somehow cause`OnBackInvokedCallback` to be invoked on a different platformview, causing an exception on Android < 13?
https://github.com/flutter/flutter/issues/151733

Apologies if this is unwanted here. I simply traced my exception back to this change and I'm seriously out of my depth here.

**Pull request merged**
